src/protocols/vnc/auth.c

```diff
@@ -31,3 +31,25 @@ char* guac_vnc_get_password(rfbClient* client) {
     return ((guac_vnc_client*) gc->data)->settings->password;
 }
 
+rfbCredential* guac_vnc_get_credentials(rfbClient* client, int credentialType) {
+    guac_client* gc = rfbClientGetClientData(client, GUAC_VNC_CLIENT_KEY);
+    rfbCredential *creds = malloc(sizeof(rfbCredential));
+    
+    if (credentialType == rfbCredentialTypeUser) {
+        creds->userCredential.username = ((guac_vnc_client*) gc->data)->settings->username;
+        creds->userCredential.password = ((guac_vnc_client*) gc->data)->settings->password;
+        return creds;
+    }
+    
+    else if (credentialType == rfbCredentialTypeX509) {
+        creds->x509Credential.x509ClientCertFile = ((guac_vnc_client*) gc->data)->settings->client_cert;
```

Comment:
Is there no way to read these keys from memory? It feels odd to have to have the certificates live within files on the guacd server, with these parameters being paths rather than PEM keys.

Reply:
I'm not sure.  The way I've implemented it, between the Client and Server parameters, is assuming that the contents of the file will be passed through, not the location of the file, but I'm not sure one way or the other.

Unfortunately, if those values represent the locations of a file, rather than the contents, I'm not sure there's much to do about it within the libVNCclient implementation, since that's internal to that implementation.

I can think of one way to work around it - we could have guacd write out the contents of the file to a location on-disk, have libvncclient read it in, then immediately delete the file.  This is a little risky, as it has the potential to expose the contents of the private key to locations where it could be read by others on the system.

I'm open to suggestions, even if it's "let's not implement this right now."

Reply:
Well, after digging into the source code, it does indeed look like libvncclient expects this to be the path to the file, not the contents of the file.  Which stinks.

https://github.com/LibVNC/libvncserver/blob/f997b5a75fa171d79c5e568b7157fba83c8d8355/libvncclient/tls_gnutls.c#L416-L425

Which calls `gnutls_certificate_set_x509_key_file()`, which takes file locations as an argument.

Aside from writing out a temporary file, are there any creative ways to point a function like `gnutls_certificate_set_x509_key_file()` to read a location in memory rather than a filesystem location?

Reply:
Well, I took a stab at writing the output to temp files.  Feels like the block of code that would be next to the word "kludge" in a programming dictionary...

Reply:
Yeah, I agree that writing credentials to temporary files on disk is not the way to go.

Perhaps the username half of this should be split away from the X.509 half? That would narrow scope to something that's definitely solvable. Supporting X.509 might need to wait for improvements to the underlying library.

Reply:
Yeah, I'm fine with not doing the X509 bits for now.  I'll go through and just pull those out, and we'll focus this on the username + password portion.